File: Coder Army/Patterns/2_FactoryDesignPattern/Javascript code/simpleDesignPattern.js
class Burger {
    prepare() {
        throw new Error('warning prepare() needs to be implemented');
    }
}

class BasicBurger extends Burger {
    prepare() {
        console.log('Preparing basic burger with bun, patty and Ketchup!');
    }
}

class StandardBurger extends Burger {
    prepare() {
        console.log('Preparing Standard burger with bun, patty, cheese, lettuce and Ketchup!');
    }
}

class PremiumBurger extends Burger {
    prepare() {
        console.log('Preparing premium burger with bun, patty, extra cheese, lettuce and Ketchup!');
    }
}

class BurgerFactory {
    createBurger(type) {
        if (type === 'basic') {
            return new BasicBurger();
        } else if (type === 'standard') {
            return new StandardBurger();
        } else if (type === 'premium') {
            return new PremiumBurger();
        } else {
            throw new Error('Not delivering this type!');
        }
    }
}

// Usage
const factory = new BurgerFactory();
const burger = factory.createBurger('basic');
burger.prepare();  

File: Coder Army/Patterns/2_FactoryDesignPattern/C++ code/simpleDesignPattern.cpp
#include <bits/stdc++.h>
using namespace std;

class Burger
{
public:
    virtual void prepare() = 0;
    virtual ~Burger() {}
};

class BasicBurger : public Burger
{
public:
    void prepare() override
    {
        cout << "Basic burger\n";
    }
};
class StandardBurger : public Burger
{
public:
    void prepare() override
    {
        cout << "Standard burger\n";
    }
};
class PremiumBurger : public Burger
{
public:
    void prepare() override
    {
        cout << "Premium burger\n";
    }
};

class SimpleFactory
{
public:
    Burger *create(const string &type)
    {
        if (type == "Basic")
        {
            return new BasicBurger();
        }
        else if (type == "Standard")
        {
            return new StandardBurger();
        }
        else if (type == "Premium")
        {
            return new PremiumBurger();
        }
        else
        {
            return nullptr;
        }
    }
};
int main()
{
    string type = "Standard";

    SimpleFactory *factory = new SimpleFactory();
    Burger *burger = factory->create(type);
    burger->prepare();
    delete factory;
    delete burger;

    return 0;
}